src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,313 @@
+            get
+            {
+                if (_index < 0)
+                    return Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).AsSpan(_index & bitMask, _length);
+                return new Span<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+            }
+        }
+
+        public unsafe MemoryHandle Retain(bool pin = false)
+        {
+            MemoryHandle memoryHandle;
+            if (pin)
+            {
+                if (_index < 0)
+                {
+                    memoryHandle = Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Pin();
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(Unsafe.As<T[]>(_arrayOrOwnedMemory), GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
+                    memoryHandle = new MemoryHandle(null, pointer, handle);
+                }
+            }
+            else
+            {
+                if (_index < 0)
+                {
+                    Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Retain();
+                    memoryHandle = new MemoryHandle(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory));
+                }
+                else
+                {
+                    memoryHandle = new MemoryHandle(null);
+                }
+            }
+            return memoryHandle;
+        }
+
+        /// <summary>
+        /// Get an array segment from the underlying memory. 
+        /// If unable to get the array segment, return false with a default array segment.
+        /// </summary>
+        public bool TryGetArray(out ArraySegment<T> arraySegment)
+        {
+            if (_index < 0)
+            {
+                if (Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).TryGetArray(out var segment))
+                {
+                    arraySegment = new ArraySegment<T>(segment.Array, segment.Offset + (_index & bitMask), _length);
+                    return true;
+                }
+            }
+            else
+            {
+                arraySegment = new ArraySegment<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+                return true;
+            }
+
+            arraySegment = default(ArraySegment<T>);
+            return false;
+        }
+
+        /// <summary>
+        /// Copies the contents of this span from the memory into a new array.  This heap
+        /// allocates, so should generally be avoided, however it is sometimes
+        /// necessary to bridge the gap with APIs written in terms of arrays.
+        /// </summary>
+        public T[] ToArray() => Span.ToArray();
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object obj)
+        {
+            if (obj is ReadOnlyMemory<T> readOnlyMemory)
+            {
+                return readOnlyMemory.Equals(this);
+            }
```

Comment:
The pattern matching `is` syntax on structs doesn't produce ideal results.  This code:
```C#
if (obj is ReadOnlyMemory<T> readOnlyMemory) return readOnlyMemory.Equals(this);
```
results in the IL:
```
        IL_0000: ldarg.1
        IL_0001: stloc.1
        IL_0002: ldloc.1
        IL_0003: isinst valuetype ReadOnlyMemory`1<!T>
        IL_0008: ldnull
        IL_0009: cgt.un
        IL_000b: dup
        IL_000c: brtrue.s IL_0019
        IL_000e: ldloca.s 2
        IL_0010: initobj valuetype ReadOnlyMemory`1<!T>
        IL_0016: ldloc.2
        IL_0017: br.s IL_001f
        IL_0019: ldloc.1
        IL_001a: unbox.any valuetype ReadOnlyMemory`1<!T>
        IL_001f: stloc.0
        IL_0020: brfalse.s IL_0031
        IL_0022: ldloca.s 0
        IL_0024: ldarg.0
        IL_0025: constrained. valuetype ReadOnlyMemory`1<!T>
        IL_002b: callvirt instance bool [mscorlib]System.Object::Equals(object)
        IL_0030: ret
```
whereas this code:
```C#
if (obj is ReadOnlyMemory<T>) return ((ReadOnlyMemory<T>)obj).Equals(this);
```
results in the IL:
```
        IL_0000: ldarg.1
        IL_0001: isinst valuetype ReadOnlyMemory`1<!T>
        IL_0006: brfalse.s IL_001e
        IL_0008: ldarg.1
        IL_0009: unbox.any valuetype ReadOnlyMemory`1<!T>
        IL_000e: stloc.0
        IL_000f: ldloca.s 0
        IL_0011: ldarg.0
        IL_0012: constrained. valuetype ReadOnlyMemory`1<!T>
        IL_0018: callvirt instance bool [mscorlib]System.Object::Equals(object)
        IL_001d: ret
```

Reply:
> doesn't produce ideal results

To clarify, it doesn't produce an incorrect result though, correct?
Are you suggesting I change the impl to this since it produces smaller IL?
```C#
if (obj is ReadOnlyMemory<T>) return ((ReadOnlyMemory<T>)obj).Equals(this);
```

Reply:
> it doesn't produce an incorrect result though, correct?

Correct

> Are you suggesting I change the impl to this since it produces smaller IL?

Yes